src/Bicep.Core/Diagnostics/ResultWithException.cs

```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using Bicep.Core.Utils;
+
+namespace Bicep.Core.Diagnostics;
+
+public class ResultWithException<TSuccess> : Result<TSuccess, Exception>
+    where TSuccess : class
+{
+    public ResultWithException(TSuccess success) : base(success) { }
+
+    public ResultWithException(Exception exception) : base(exception) { }
+}
```

Comment:
Can you elaborate on the need for this new type vs reusing `ResultWithDiagnostics`?

Reply:
I'm not dealing with diagnostics, no need to complicate it with that concept.

I don't actually need an Exception, just a message, but Exception provides a way to separate out the cases of succeed vs failed because not sources vs failed because of an error retrieving/parsing sources (by having the two failure cases be differentiated based on Exception type).  

Originally had this as a new record SourceArchiveResult where the success case could have a null result, but that seems unnecessary given Result is intended for this.  But Result<T,U> doesn't currently allow T to be nullable, thus using SourcesNotAvailableException for that scenario.

Reply:
> But Result<T,U> doesn't currently allow T to be nullable, thus using SourcesNotAvailableException for that scenario.

I removed that restriction in #12956, but it hasn't been merged. Feel free to incorporate those changes into this branch if it helps simplify things.

Reply:
Cool, thanks.  I considered doing that myself, but had mixed feelings about whether it should best be represented by a null success or a specific exception.  Since you've done it I might refactor later.  But probably not. :-)